src/lc3/vm.rs

```diff
@@ -257,29 +208,110 @@ impl VM {
+        self.set_register(r0, value_read)?;
+        self.update_flags(r0)?;
+        Ok(())
+    }
+
+    fn ldr(&mut self, instr: u16) -> Result<(), VMError> {
+        let r0 = (instr >> 9) & 0b0111;
+        let r1 = (instr >> 6) & 0b0111;
+        let value_in_r1 = *self.get_register(r1)?;
+        let pc_offset = sign_extend(instr & 0b0011_1111, 6)?;
+        let address = value_in_r1
+            .checked_add(pc_offset)
+            .ok_or(VMError::MemoryIndex(String::from(
+                "Overflow with offset in Load Register",
+            )))?;
+        let value_read = self.mem_read(address.into())?;
+        self.set_register(r0, value_read)?;
+        self.update_flags(r0)?;
+        Ok(())
+    }
+
+    fn ldi(&mut self, instr: u16) -> Result<(), VMError> {
+        let r0 = (instr >> 9) & 0b0111;
+        let pc_offset = sign_extend(instr & 0b0001_1111_1111, 9)?;
+        let address = self
+            .pc
+            .checked_add(pc_offset)
+            .ok_or(VMError::MemoryIndex(String::from(
+                "Overflow with offset in Load Indirect",
+            )))?;
+        let value_read = self.mem_read(address.into())?;
+        self.set_register(r0, self.mem_read(value_read.into())?)?;
+        self.update_flags(r0)?;
+        Ok(())
+    }
+
+    fn st(&mut self, instr: u16) -> Result<(), VMError> {
+        let r0 = (instr >> 9) & 0b0111;
+        let value_in_r0 = *self.get_register(r0)?;
+        let pc_offset = sign_extend(instr & 0b0001_1111_1111, 9)?;
+        let address = self
+            .pc
+            .checked_add(pc_offset)
+            .ok_or(VMError::MemoryIndex(String::from(
+                "Overflow with offset in Store",
+            )))?;
+        self.mem_write(value_in_r0, address.into())?;
+        Ok(())
+    }
+
+    fn sti(&mut self, instr: u16) -> Result<(), VMError> {
+        let r0 = (instr >> 9) & 0b0111;
+        let value_in_r0 = *self.get_register(r0)?;
+        let pc_offset = sign_extend(instr & 0b0001_1111_1111, 9)?;
+        let address = self
+            .pc
+            .checked_add(pc_offset)
+            .ok_or(VMError::MemoryIndex(String::from(
+                "Overflow with offset in Store Indirect",
+            )))?;
+        let value_read = self.mem_read(address.into())?;
+        self.mem_write(value_in_r0, value_read.into())?;
+        Ok(())
+    }
+
+    fn str(&mut self, instr: u16) -> Result<(), VMError> {
+        let r0 = (instr >> 9) & 0b0111;
+        let value_in_r0 = *self.get_register(r0)?;
+        let r1 = (instr >> 6) & 0b0111;
+        let value_in_r1 = *self.get_register(r1)?;
+        let pc_offset = sign_extend(instr & 0b0011_1111, 6)?;
+        let address = value_in_r1
+            .checked_add(pc_offset)
+            .ok_or(VMError::MemoryIndex(String::from(
+                "Overflow with offset in Store Register",
+            )))?;
+        self.mem_write(value_in_r0, address.into())?;
+        Ok(())
```

Comment:
It would be nice if we put instructions in a separated file as you did with byte operations. Also this will be much easier to review when we update something.

Reply:
I agree it would be better for reviewing, but it is part of the VM and separating it into files would make me made everything public. Is it worth it?

Reply:
We can move the instruction to the opcode file

Reply:
Done, [650e5e3](https://github.com/tomip01/lc3-vm/pull/3/commits/650e5e347e24cc89f8d7d7f936ba1cd63babe970)